### PR TITLE
Issue #15457: validate our website javadoc links

### PR DESCRIPTION
**Pull request by rnveach**
Issue #15457

### CODE REVIEW

**Comment by rnveach** [comment]
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/27191/workflows/9d488a74-9a47-4bd5-845e-518afa737c5e/jobs/660275
> `<a href="apidocs/com/puppycrawl/tools/checkstyle/checks/styleguides/google-java-style-20220203/javaguide.html#s4.8.6.1-block-comment-style">../styleguides/google-java-style-20220203/javaguide.html#s4.8.6.1-block-comment-style</a>: doesn't exist.`

It appears to be working. Directory needs to be corrected a bit more. This can't tell the difference between the href and the text of the href. I am not sure if we always ensure we write `href="https" on a single line.

**Comment by rnveach** [comment]
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/27192/workflows/beb9b029-8adc-426e-b7b5-d7326fc061b6/jobs/660302
I would need to look at what this one is about.

But I do see an issue with this I should have seen coming. We need replace a dynamic number of sub-directories and replace with `../` as many times as it takes to get out of javadoc folder. It won't always be a fixed count.

**Comment by rnveach** [comment]
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/27199/workflows/c6f1d1a9-1319-41aa-b14a-8426e000f10b/jobs/660597
linkcheck is passing.

I can confirm locally that the false positive link is gone.

**Comment by rnveach** [comment]
@romani @nrmancuso Are we ok with this resolving https://github.com/checkstyle/checkstyle/issues/15457 ? If so, then I will fix the PR up. Please note I am weak in bash and had to use ChatGPT to help with the dynamic directory depth replacement.

File: .ci/run-link-check-plugin.sh [hunk 1]
```diff
@@ -7,8 +7,34 @@ pwd
...
+}
+
+# convert all javadoc website links to relative links
+grep -rl "https://checkstyle.org" src/main/java | while read -r file; do
+    replacement=$(calculate_replacement "$file")
+    sed -i "s|https://checkstyle.org|$replacement|g" "$file"
+done
+
+mvn -e --no-transfer-progress -Pno-validations clean site -Dlinkcheck.skip=false \
+   -Dmaven.javadoc.skip=false
```
Comment (rnveach) [inline .ci/run-link-check-plugin.sh:33]: This command was replaced using `no-validations` so we don't have to manually disable checks, especially if one is added, and we can clearly see what is being enabled.

**Comment by romani** [comment]
>I am not sure if we always ensure we write `href="https" on a single line.

we almost always do this, links are not wrap-able.

**Comment by romani** [comment]
>On my local, error count went from 1362 to 1360.

My worries it that we do very complicated code changes just to hit 2 violations out of  1362 .
This code might give us more problems on edgecases  in future, and benefit of it not significant.
And workaround for this problems with 2 links is too make release, that is temporal problem of summer activity.

File: .ci/run-link-check-plugin.sh (hunk 1 above)
Comment (romani) [reply .ci/run-link-check-plugin.sh:33]: I think we can make sepatate PR for this update only and merge it quickly.

File: .ci/run-link-check-plugin.sh [hunk 2]
```diff
@@ -7,8 +7,34 @@ pwd
...
+
+# revert all modified javadoc files for post validation
+git checkout HEAD -- src/main/java
```
Comment (romani) [inline .ci/run-link-check-plugin.sh:36]: this is a concern, I usually always treated CI tasks or any validation task as commands that ready to be executed at any time in root of repo.
this update will revert all user uncommited changes, not good.
if we agreed to move forward with this links changes we need to copy checkouted repo to .ci-temp and do all code changes in that folder, and this checkout might not be required at all.

**Review by romani — CHANGES_REQUESTED** [review]
@nrmancuso, please share your thoughts on issue in general.

for post approval time:

**Comment by rnveach** [comment]
> My worries it that we do very complicated code changes just to hit 2 violations out of 1362 .

It is misleading to say this is to hit 2 instances.

This is to hit any changes in master which has not been released. This **does** validate other links properly, by pointing them to master instead of the real server, but those will work with real server **ONLY WHEN** they match master currently.

This release it is only these 2 false positives. If we added new grammar token, it would impact those too, as all our tokens point to javadoc URL. Most likely most releases would not be impacted by this.

This also provides us a green CI in master so it does not confuse people with a false positive failure.

If we do not want this fix, then the only other option is to ignore all links directly to our website URL, and we will validate them after the release and fix them then. This means we will always be fixing failures of the previous release in the next release's snapshot. This is basically what we are doing now by ignoring master CI failure, unless we rigorously recheck the failure every commit to very it is false and not a positive.

File: .ci/run-link-check-plugin.sh (hunk 1 above)
Comment (nrmancuso) [reply .ci/run-link-check-plugin.sh:33]: Yeah, this is good

**Review by nrmancuso — COMMENTED** [review]
Maybe this seems weird, but I will propose another approach:
1. Run linkcheck once
2. Modify the linkcheck report instead of the javadocs
3. Run linkcheck again without `clean` (or use this as a cached file)

I am not sure if this would work, but it is probably easier/better than modifying java files IMO

**Comment by romani** [comment]
I like idea of updates in generated website. It will be safer and we can hack anything we want by any craziest bash/.... code.

**Pull request closed without merging**